xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,231 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Kodi; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonManager.h"
+#include "BinaryAddonBase.h"
+
+#include "addons/AddonManager.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+using namespace ADDON;
+
+CBinaryAddonManager::~CBinaryAddonManager()
+{
+  CAddonMgr::GetInstance().Events().Unsubscribe(this);
+}
+
+bool CBinaryAddonManager::Init()
+{
+  CAddonMgr::GetInstance().Events().Subscribe(this, &CBinaryAddonManager::OnEvent);
+
+  BINARY_ADDON_LIST binaryAddonList;
+  if (!CAddonMgr::GetInstance().GetInstalledBinaryAddons(binaryAddonList))
+  {
+    CLog::Log(LOGNOTICE, "CBinaryAddonManager::%s: No binary addons present and related manager, init not necessary", __FUNCTION__);
+    return true;
+  }
+
+  for (auto addon : binaryAddonList)
+    AddAddonBaseEntry(addon);
+
+  return true;
+}
+
+bool CBinaryAddonManager::HasInstalledAddons(const TYPE &type) const
+{
+  CSingleLock lock(m_critSection);
+  for (auto info : m_installedAddons)
+  {
+    if (info.second->IsType(type))
+      return true;
+  }
+  return false;
+}
+
+bool CBinaryAddonManager::HasEnabledAddons(const TYPE &type) const
+{
+  CSingleLock lock(m_critSection);
+  for (auto info : m_enabledAddons)
+  {
+    if (info.second->IsType(type))
+      return true;
+  }
+  return false;
+}
+
+void CBinaryAddonManager::GetAddonInfos(BinaryAddonBaseList& addonInfos, bool enabledOnly, const TYPE &type) const
+{
+  CSingleLock lock(m_critSection);
+
+  const BinaryAddonMgrBaseList* addons;
+  if (enabledOnly)
+    addons = &m_enabledAddons;
+  else
+    addons = &m_installedAddons;
+
+  for (auto info : *addons)
+  {
+    if (type == ADDON_UNKNOWN || info.second->IsType(type))
+    {
+      addonInfos.push_back(info.second);
+    }
+  }
+}
+
+const BinaryAddonBasePtr CBinaryAddonManager::GetInstalledAddonInfo(const std::string& addonId, const TYPE &type/* = ADDON_UNKNOWN*/) const
+{
+  CSingleLock lock(m_critSection);
+
+  auto addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end() && (type == ADDON_UNKNOWN || addon->second->IsType(type)))
+    return addon->second;
+
+  CLog::Log(LOGERROR, "CBinaryAddonManager::%s: Requested addon '%s' unknown as binary", __FUNCTION__, addonId.c_str());
+  return nullptr;
+}
+
+AddonPtr CBinaryAddonManager::GetRunningAddon(const std::string& addonId) const
+{
+  CSingleLock lock(m_critSection);
+
+  auto addon = m_installedAddons.find(addonId);
+  if (addon != m_installedAddons.end())
+    return addon->second->GetActiveAddon();
+
+  return nullptr;
+}
+
+bool CBinaryAddonManager::AddAddonBaseEntry(BINARY_ADDON_LIST_ENTRY& entry)
+{
+  BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>(entry.second);
+  if (!base->Create())
```

Comment:
Holding the internal lock while doing io and external calls like this is a recipe for deadlocks. I don't have an easy fix for this, but it's better to write initialization in such a way that it doesn't require calls to be made in-between reading and writing to m_installedAddons/m_enabledAddons